# **Maestría en Inteligencia Artificial Aplicada**

## **Curso: Inteligencia Artificial y Aprendizaje Automático**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

## **Actividad de las Semanas 5 y 6**
### **Problema de asignación de créditos: South German Dataset.**


## **Nombres y matrículas:**


* ### ALICIA YOVANNA CANTA PANDAL - A01796095

* ### HIRAM GARCIA AUSTRIA - A00378771

* ### ALFREDO RIGOBERTO ALVAREZ SUAREZ - A01796142



# **Parte I: Partición, análisis y pre-procesamiento de los datos.**

## **Ejercicio 1**

In [144]:
# Aquí deberás incluir todas las librerías que requieras durante esta actividad:
import numpy as np
import pandas as pd

from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC
from sklearn.model_selection import RepeatedStratifiedKFold, train_test_split, cross_validate, GridSearchCV
from imblearn.metrics import geometric_mean_score
from imblearn.over_sampling import KMeansSMOTE
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import make_scorer

# Bosque aleatorio
from sklearn.ensemble import RandomForestClassifier

In [145]:
# Si se desean comentar algunos de los Warnings.
import warnings
warnings.filterwarnings('ignore')

In [146]:
# Carga y renombra los nombres de las columnas del alemán al inglés y desplegamos
# de nuevo el DataFrame para ver el resultado obtenido:

# ************* Incluye aquí tu código:*****************************

df = pd.read_csv('./SouthGermanCredit.asc', sep=' ')
df.columns = [
    'status',
    'duration',
    'credit_history',
    'purpose',
    'amount',
    'savings',
    'employment_duration',
    'installment_rate',
    'personal_status_sex',
    'other_debtors',
    'present_residence',
    'property',
    'age',
    'other_installment_plans',
    'housing',
    'number_credits',
    'job',
    'people_liable',
    'telephone',
    'foreign_worker',
    'credit_risk'
]

# *********** Aquí termina la sección de agregar código *************

df.head().T

0     1    2     3     4
status                      1     1    2     1     1
duration                   18     9   12    12    12
credit_history              4     4    2     4     4
purpose                     2     0    9     0     0
amount                   1049  2799  841  2122  2171
savings                     1     1    2     1     1
employment_duration         2     3    4     3     3
installment_rate            4     2    2     3     4
personal_status_sex         2     3    2     3     3
other_debtors               1     1    1     1     1
present_residence           4     2    4     2     4
property                    2     1    1     1     2
age                        21    36   23    39    38
other_installment_plans     3     3    3     3     1
housing                     1     1    1     1     2
number_credits              1     2    1     2     2
job                         3     3    2     2     2
people_liable               2     1    2     1     2
telephone                   1     1    1     1     1
foreign_worker              2     2    2     1     1
credit_risk                 1     1    1     1     1

## **Ejercicio 2**

In [147]:
# Realiza a continuación una transformación para que la clase negativa (buen cliente)
# quede con el valor de 0 y la clase positiva (mal cliente) quede con el valor de 1.

# ************* Incluye aquí tu código:*****************************

df['credit_risk'] = df['credit_risk'].replace({'1': 0, '0': 1})

# *********** Aquí termina la sección de agregar código *************


print(df['credit_risk'].value_counts())

credit_risk
1    700
0    300
Name: count, dtype: int64


## **Ejercicio 3**

In [148]:
# Realiza una partición solicitada de entrenamiento y prueba.
# Los nombres de los conjuntos deberán ser como se indican en los print de abajo:

# ************* Incluye aquí tu código:*****************************

X = df.drop('credit_risk', axis=1)
y = df['credit_risk']

# Mismos porcentajes que se utilizaron en el artículo: 70% - 30%
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# *********** Aquí termina la sección de agregar código *************


# Mostremos las dimensiones de la partición generada:
print("Dimensiones:")
print("Entrenamiento:", Xtrain.shape, ytrain.shape)
print("Prueba:", Xtest.shape, ytest.shape)

# Y el porcentaje de cada clase de la variable de salida:
tmp = ytrain.sum()/ytrain.shape[0]
print("\nPorcentaje clases Positiva:%.2f%%, y Negativa:%.2f%%" % (100*(1-tmp),tmp*100))


Dimensiones:
Entrenamiento: (700, 20) (700,)
Prueba: (300, 20) (300,)

Porcentaje clases Positiva:30.00%, y Negativa:70.00%


### **Con base al porcentaje de los niveles de la variable de salida ¿podemos decir que tenemos un problema de datos desbalanceado? ¿Por qué?**



### ++++++++ Inicia la sección de agregar texto: +++++++++++

Tenemos un problema de clases desbalanceadas ya que la clase positiva (mal cliente) es minoritaria (30%), mientras que la clase negativa (buen cliente) es mayoritaria (70%).

Para resolver este problema podemos realizar un balanceo de las clases en el conjunto de entrenamiento con técnicas de sobre-muestreo en la clase minoritaria, o bien técnicas de sub-muestreo en la clase mayoritaria.

### ++++++++ Termina la sección de agregar texto: +++++++++++



## **Ejercicio 4**

In [149]:
# De acuerdo a la información de la Tabla 3 del artículo de la IEEE
# define las variables correspondientes que se indican a continuación:

# ************* Incluye aquí tu código:*****************************

# Variables numéricas:
lista_paper_num = [
    'duration',
    'amount',
    'age',
    'people_liable'
]

# Variables ordinales:
lista_paper_ord = [
    'employment_duration',
    'installment_rate',
    'present_residence',
    'property',
    'number_credits',
    'job'
]

# Variables nominales:
lista_paper_cat = [
    'status',
    'credit_history',
    'purpose',
    'savings',
    'personal_status_sex',
    'other_debtors',
    'other_installment_plans',
    'housing',
    'telephone',
    'foreign_worker',
]


# *********** Aquí termina la sección de agregar código *************


## **Ejercicio 5**

In [150]:
# Transformaciones que se aplicarán a las variables numéricas usando la clase Pipeline de sklearn:

# ************* Incluye aquí tu código:*****************************

# Variables numéricas:
numericas_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', MinMaxScaler()) # aplicando la misma normalización que el artículo
])
numericas_pipe_nombres = lista_paper_num


# Variables categóricas-Nominales:
nominales_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', drop='first'))
])
nominales_pipe_nombres = lista_paper_cat

# Variables categóricas-ordinales:
ordinales_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])
ordinales_pipe_nombres = lista_paper_ord

# Conjuntas las transformaciones de todo tipo de variable y
# deja sin procesar aquellas que hayas decidido no transformar:

columnasTransformer = ColumnTransformer([
    ('numericas', numericas_pipe, numericas_pipe_nombres),
    ('nominales', nominales_pipe, nominales_pipe_nombres),
    ('ordinales', ordinales_pipe, ordinales_pipe_nombres)
], remainder='passthrough')

# *********** Aquí termina la sección de agregar código *************


## **Ejercicio 6**

In [151]:
# Como se va a utilizar Validación-Cruzada, concatena los conjuntos de entrenamiento y prueba
# en uno nuevo conjunto aumentado que llamaremos trainval para utilizar como entrenamiento:


# ************* Incluye aquí tu código:**************************

Xtraintest = pd.concat([Xtrain, Xtest], axis=0)
ytraintest = pd.concat([ytrain, ytest], axis=0)

# *********** Aquí termina la sección de agregar código *************


# Veamos cuántas variables nuevas se introducen con las transformaciones One-Hot-Encoding:
Xtmp = Xtraintest.copy()
tmp = columnasTransformer.fit_transform(Xtmp)
print("Dimensión de las variables de entrada ANTES de las transformaciones:", Xtmp.shape)
print("Dimensión de las variables de entrada DESPUÉS de las transformaciones:", tmp.shape)

Dimensión de las variables de entrada ANTES de las transformaciones: (1000, 20)
Dimensión de las variables de entrada DESPUÉS de las transformaciones: (1000, 41)


## **Ejercicio 7**

In [152]:
# Definimos a continuación la función que llamamos "mis_modelos" que incluye
# todos los modelos que deseamos comparar en el ejercicio.


def mis_modelos():
  modelos, nombres = list(), list()


  # ************* Incluye aquí tu código:**************************
  #
  # Deberás incluir en cada modelo los argumentos que consideres
  # adecuados para que cada uno converja y no esté sobre-entrenado
  # con respecto a la métrica de la exactitud (accuracy).


  # Regresión Logística - Logistic Regression-LR:
  # https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

  # Este valor de "penalty" no se debe de cambiar, ya que define el modelo sin regularización.
  modelos.append(LogisticRegression(penalty=None, solver='lbfgs', max_iter=2000, random_state=1))
  nombres.append('LR')




  # k-Vecinos más Cercanos : k-Nearest-Neighbors-kNN:
  # https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html

  modelos.append(KNeighborsClassifier())
  nombres.append('kNN')




  # Árbol de decisiones-DecisionTree-DT:
  # https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html

  modelos.append(DecisionTreeClassifier(max_depth = None,       # None, 5, 3, 1
                                        min_samples_split=2,   # 2,3,5, 20
                                        #min_samples_leaf=1,  # trata al inicio de usar solo uno de estos, split o leaf, para su mejor comprensión.
                                        #random_state=7
                                        ))
  nombres.append('DTree')




  # Bosque Aleatorio-RandomForest-RF:
  # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html

  modelos.append(RandomForestClassifier(max_depth=None, min_samples_split=2, n_estimators=300))
  nombres.append('RF')




  # XGBoosting:
  # https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.XGBClassifier
  # https://xgboost.readthedocs.io/en/stable/parameter.html

  modelos.append(XGBClassifier(booster='gbtree', learning_rate=0.3, max_depth=3, n_estimators=50, subsample=1.0, objective='binary:logistic', n_jobs=-1))
  nombres.append('XGBoost')




  # Red neuronal de Perceptrón Multicapa-MLP:
  # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html

  modelos.append(MLPClassifier(activation='logistic', alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=100))
  nombres.append('MLP')




  # Máquina de Vectores de Soporte-SVM:
  # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

  modelos.append(SVC(kernel='rbf',
                     C= 1.0,      # 1.0, 0.00001, 1000           # inversamente proporcional a la constante de regularización L2.
                     gamma= 'scale',           # scale,  0.005
                     #class_weight='balanced',     # Siempre puedes hacer uso del balanceo en caso de que ayude.
                     #random_state=7
                     ))
  nombres.append('SVM')

  return modelos, nombres


# Técnica de sub muestreo (under sampling) y/o sobre muestreo (oversampling) utilizada:

# Aplicando sub-muestreo con Near Miss:
# near_miss = NearMiss(version=1, n_neighbors=3)
# Xtv_uo, ytv_uo = near_miss.fit_resample(Xtraintest, ytraintest)

# Aplicando sobre-muestreo con K-Means SMOTE:
kmeans_smote = KMeansSMOTE(random_state=42)
Xtv_uo, ytv_uo = kmeans_smote.fit_resample(Xtraintest, ytraintest)

# Aplicando sobre-muestreo con SMOTE:
# mi_uoSampling = SMOTE(random_state=42)
# Xtv_uo, ytv_uo = mi_uoSampling.fit_resample(Xtraintest, ytraintest)


# *********** Aquí termina la sección de agregar código *******************





# Entrenamos cada uno de los modelos y desplegamos la métricas de Train y Val.

# NOTA: Observa que el método de Validación-Cruzada llama  a los resultados
#       de "validation" como "test":

modelos, nombres = mis_modelos()
resultados = list()

for i in range(len(modelos)):

  # Definimos nuestro pipeline con las transformaciones y los modelos:
  pipeline = Pipeline(steps=[('ct',columnasTransformer),('m',modelos[i])])

  # Aplicaremos validación-cruzada:
  micv = RepeatedStratifiedKFold(n_splits=5,
                                 n_repeats=3,
                                 random_state=5     # agreguemos una semilla para estabilizar resultados.
                                 )


  # Definimos las métricas que desamos recuperar:
  mismetricas = {'miaccuracy':'accuracy','miprecision':'precision','mirecall':'recall',
                 'mifi':'f1','miauc':'roc_auc','migmean':make_scorer(geometric_mean_score)}

  # Llevamos a cabo el entrenamiento:
  scores = cross_validate(pipeline,
                          Xtv_uo,
                          ytv_uo,
                          scoring=mismetricas,
                          cv=micv,
                          return_train_score=True,
                          )

  # Guardemos el resultado de cada modelo para análisis posteriores.
  resultados.append(scores)

  # Desplegamos los valores de las métricas para verificar si no hay
  # subentrenamiento o sobreentrenamiento:
  print('>> %s' % nombres[i])
  for j,k in enumerate(list(scores.keys())):
    if j>1:
      print('\t %s %.3f (%.3f)' % (k, np.mean(scores[k]),np.std(scores[k])))

>> LR
	 test_miaccuracy 0.815 (0.023)
	 train_miaccuracy 0.831 (0.004)
	 test_miprecision 0.807 (0.026)
	 train_miprecision 0.821 (0.004)
	 test_mirecall 0.827 (0.036)
	 train_mirecall 0.846 (0.008)
	 test_mifi 0.817 (0.024)
	 train_mifi 0.833 (0.005)
	 test_miauc 0.904 (0.022)
	 train_miauc 0.922 (0.005)
	 test_migmean 0.814 (0.023)
	 train_migmean 0.831 (0.004)
>> kNN
	 test_miaccuracy 0.784 (0.027)
	 train_miaccuracy 0.851 (0.008)
	 test_miprecision 0.773 (0.033)
	 train_miprecision 0.834 (0.013)
	 test_mirecall 0.805 (0.038)
	 train_mirecall 0.877 (0.008)
	 test_mifi 0.788 (0.026)
	 train_mifi 0.855 (0.008)
	 test_miauc 0.861 (0.032)
	 train_miauc 0.936 (0.006)
	 test_migmean 0.783 (0.027)
	 train_migmean 0.851 (0.009)
>> DTree
	 test_miaccuracy 0.773 (0.019)
	 train_miaccuracy 1.000 (0.000)
	 test_miprecision 0.779 (0.026)
	 train_miprecision 1.000 (0.000)
	 test_mirecall 0.763 (0.025)
	 train_mirecall 1.000 (0.000)
	 test_mifi 0.770 (0.019)
	 train_mifi 1.000 (0.000)
	 test_miauc

In [153]:
# Creando una validación cruzada para los 3 mejores modelos

############################ Red neuronal de Perceptrón Multicapa-MLP ############################
mlp = MLPClassifier(random_state=1)

# Parámetros a probar
params_mlp = {
    'hidden_layer_sizes': [(6,), (15, 15), (30,), (50, 50), (100,)], # número de neuronas en cada capa oculta.
    'activation': ['logistic'],
    'max_iter': [50, 100, 200, 500, 1000],
    'alpha': [0.0001], # término de regularización L2.
}

# Validación cruzada
grid_search_mlp = GridSearchCV(estimator=mlp, param_grid=params_mlp, cv=5, scoring='accuracy')

# Entrenamiento
grid_search_mlp.fit(Xtv_uo, ytv_uo)

# Resultados en forma de lista
print(f'Mejores parámetros MLP: {grid_search_mlp.best_params_}')
print(f'Mejor accuracy MLP: {grid_search_mlp.best_score_}')
print('\n')

############################ Bosque Aleatorio (RandomForest - RF) ############################
rf = RandomForestClassifier(random_state=1)

# Parámetros a probar
params_rf = {
    'n_estimators': [50, 100, 200, 300],
    'max_depth': [None, 1, 2, 3, 4, 5, 6],
    'min_samples_split': [2, 5, 15],
}

# Validación cruzada
grid_search_rf = GridSearchCV(estimator=rf, param_grid=params_rf, cv=5, scoring='accuracy')

# Entrenamiento
grid_search_rf.fit(Xtv_uo, ytv_uo)

# Resultados en forma de lista
print(f'Mejores parámetros RF: {grid_search_rf.best_params_}')
print(f'Mejor accuracy RF: {grid_search_rf.best_score_}')
print('\n')


############################ XGBoosting ############################
xgb = XGBClassifier(random_state=1)

# Parámetros a probar
params_xgb = {
    'booster': ['gbtree'],
    'n_estimators': [50, 100, 200, 300],
    'max_depth': [1, 3, 6],
    'learning_rate': [0.3, 0.01, 0.000001],
    'subsample': [1.0, 0.9, 0.8, 0.5],
    'objective': ['binary:logistic'],
    'n_jobs': [-1]
}

# Validación cruzada
grid_search_xgb = GridSearchCV(estimator=xgb, param_grid=params_xgb, cv=5, scoring='accuracy')

# Entrenamiento
grid_search_xgb.fit(Xtv_uo, ytv_uo)

# Resultados en forma de lista
print(f'Mejores parámetros XGB: {grid_search_xgb.best_params_}')
print(f'Mejor accuracy XGB: {grid_search_xgb.best_score_}')
print('\n')

Mejores parámetros MLP: {'activation': 'logistic', 'alpha': 0.0001, 'hidden_layer_sizes': (50, 50), 'max_iter': 100}
Mejor accuracy MLP: 0.7817488561260804


Mejores parámetros RF: {'max_depth': None, 'min_samples_split': 2, 'n_estimators': 300}
Mejor accuracy RF: 0.7960040671072699


Mejores parámetros XGB: {'booster': 'gbtree', 'learning_rate': 0.3, 'max_depth': 3, 'n_estimators': 50, 'n_jobs': -1, 'objective': 'binary:logistic', 'subsample': 1.0}
Mejor accuracy XGB: 0.8073945094051856




## **Ejercicio 8**

### **Escribe tus conclusiones finales de la actividad.**

### ++++++++ Inicia la sección de agregar texto: +++++++++++



## **Conclusiones sobre los resultados obtenidos**

Esta actividad nos ha ayudado a entender de una mejor manera, la forma en la que se prepara un modelo de machine learning y sobre todo los ajustes que se deben hacer para obtener el mejor rendimiento posible.

Cabe mencionar que dentro de nuestros muchos ajustes, obtuvimos dos tipos de resultados:
* A) Métricas altas, pero con sobre-entrenamiento, como el caso de XGBoost y Random Forest (RF)
* B) Métricas aceptables (similares a las del reporte de IEEE), sin sobre-entrenamiento, como LogisticRegression (LR), Red neuronal de Perceptrón Multicapa (MLP), y Maquina de Vectores (SVM).

Es importante mencionar que para el siguiente análisis se siguieron las instrucciones del PDF de esta actividad, asi como el [artículo IEEE](https://ieeexplore.ieee.org/document/9239944) para la selección de las técnicas de sub-muestreo y sobre-muestreo.

## Optimización de Hiperparámetros
La forma en la que obtuvimos los mejores hiperparámetros fue haciendo una comparativa exhaustiva mediante la ayuda de la función GridSearchCV sobre los 3 modelos que obtuvieron las métricas mas altas en la primer corrida del código.

## Sub-muestreo y Sobre-muestreo
Por tratarse de un caso con clases desbalanceadas era imperativo que hiciéramos uso de una o varias técnicas de sub-muestreo y/o sobre-muestreo. Dicho esto, probamos varias técnicas tanto en cadena como por separado, tales como: NearMiss, SMOTE, SMOTETomek y KMeansSMOTE. Finalmente la que mejor resultados arrojó fue la de **KMeansSMOTE** en solitario, que es precisamente la que mejores resultados dió en el artículo de IEEE.

## Resultados finales
A continuación presentamos los resultados obtenidos en el apartado de Exactitud (Accuracy), y mas importante aun la media geométrica (Gmean), pues es el valor que toma en cuenta el desbalanceo de clases.

### MODELOS CON MAYOR PUNTUACIÓN

**Random Forest (RF)**
* test_miaccuracy 0.825 (0.026)
* train_miaccuracy 1.000 (0.000)
* test_migmean 0.820 (0.028)
* train_migmean 1.000 (0.000)

**XGBoost**
* test_miaccuracy 0.818 (0.023)
* train_miaccuracy 0.925 (0.006)
* test_migmean 0.816 (0.025)
* train_migmean 0.924 (0.006)

Ambos modelos a pesar de tener las puntuaciones mas altas, sufren de sobre-entrenamiento, ya que la diferencia entre las pruebas y el entrenamiento es mayor a un 5%. Por lo que los descartamos como posibles opciones.


### MODELOS CON MEJOR PUNTUACIÓN (sin sobre-entrenamiento)

**Logistic Regression (LR)**
* test_miaccuracy 0.815 (0.023)
* train_miaccuracy 0.831 (0.004)
* test_migmean 0.814 (0.023)
* train_migmean 0.831 (0.004)

**Maquina de Vectores (SVM)**
* test_miaccuracy 0.816 (0.021)
* train_miaccuracy 0.850 (0.005)
* test_migmean 0.812 (0.022)
* train_migmean 0.847 (0.005)

**Red neuronal de Perceptrón Multicapa (MLP)**
* test_miaccuracy 0.812 (0.024)
* train_miaccuracy 0.833 (0.006)
* test_migmean 0.811 (0.024)
* train_migmean 0.833 (0.007)

De estos 3 últimos modelos, el que tiene la mejor exactitud es SVM con 81.6%, pero la métrica que realmente considera el desbalanceo de clases es ela media geométrica y el modelo mejor puntuado es La Regresión Logística (LR) con 81.4%.


De esta forma podemos calificar como el mejor modelo de nuestro análisis a **Regresión Logística (LR)**, con un rendimiento similar al del [articulo IEEE](https://ieeexplore.ieee.org/document/9239944), y aun mas importante sin sufrir de sobre-entrenamiento, pues la diferencia entre las pruebas y el entrenamiento es de tan solo el 1.3%.


### ++++++++ Termina la sección de agregar texto. +++++++++++


>> ### **Fin de la Actividad de las Semanas 5 y 6.**